index.js

```diff
@@ -180,7 +180,7 @@ module.exports = {
     'prefer-spread': 2,
     'quotes': [2, 'single', 'avoid-escape'],
     'radix': 2,
-    'require-await': 0,
+    'require-await': 2,
```

Comment:
I'm not sure I agree with this rule. For example, our GraphQL codebase marks all async functions with `async`, even if we don't use `await` in them, simply as a way of making clear to developers that it may return a `Promise`

Reply:
I also don't think this works, though not because it conflicts with current GraphQL style as @MatthewHerbst suggests (if that were the case, we'd just disable the rule in that codebase).

Rather, I agree with the goal, but don't think this accomplishes it.

Allow me to expand on your examples a bit:
```javascript
// this function passes the rule
async function doFirstThing() {
  console.log('Start first thing');
  await console.log('Start second thing');
  console.log('Start third thing');
}

// this function passes as well
async function doSecondThing() {
  await console.log('Start fourth thing');
}
```

So what happens when we run `executeAllThings()`, as you've defined it?
```
Call stack 1:
  - doFirstThing line 1: no await, execute console.log('Start first thing');
  - doFirstThing line 2: *with* await, yield execution until this promise resolves
  - doFirstThing line 3: not executed at all on this stack
  - doSecondThing: *with* await, enqueue function block
  - execute console.log('Started all async tasks');
Call stack 2:
  - execute doSecondThing
  - execution of doFirstThing resumes at line 3
```

So you've set up `doFirstThing` line 2 and `doSecondThing` in a race to resolve. The output is:
```
Start first thing
Start second thing
Start fourth thing
Started all async tasks
Starts third thing
```

I don't think that's better. It certainly doesn't seem intuitive that things would execute in this order.

Again, if the goal is as you describe, I believe what you want is a different rule. Namely, something that would ensure all calls to an `async` function are `await`ed or themselves wrapped in a `Promise`. i.e. if we change the definition of `executeAllThings` to:
```javascript
async function executeAllThings() {
  const promise1 = await doFirstThing();
  const promise2 = await doSecondThing();
  console.log('Started all async tasks');
}
```

then we get an output much more in line with what we might expect:
```
Starts first thing // this executed synchronously
 starts second thing // starting here, all following calls are async
starts third thing
starts fourth thing
Started all async tasks
```

Even removing the `await` from one of the `do(First|Second)Thing` methods, thus making it invalid according to this rule, we get the same results.

If we don't care about the execution order of `doFirstThing` vs. `doSecondThing`, we can wrap them in a `Promise.all` and `await` that, but still have predictability over how it relates to the synchronous code in the function.

*tl;dr*
Essentially, this change suggests that missing `await` inside an `async` function is indicative of a mistake; I assert that missing `await` (or a `Promise` wrapper when _calling_ an `async` function is where the mistakes lie.

Reply:
You're not wrong about the execution order, however the synchronization of the promises is not in scope for this change. As you said if you are concerned about synchronization you would use a `Promise.all` or another method that can handle the synchronizing of promises.

However the point is that if you omit `await` in the `doFirstThing()` you get the behaviour that is completely incongruent with when the final/resolved value from an `async` method is returned. In this case, it is not about the initial execution, in fact you could still have a function with `await` inside of it that looks like
```
async doFirstThing() {
  console.log('Start first thing');
  const output = await someOtherThing();
  return other thing;
}
```
And then the call stack would still look like you described above **but** the resulting call would be guaranteed to callback only after the next callstack executes. Therefore maintaining consistency of execution among all `async` functions. 

If there is never *any* `await` used inside the function block you will return the result of the function synchronously, and in doing-so may create synchronization issues that are implicitly hard to reason about when debugging - because the functions using `async` ought to be run **later** which means at least one call-stack in the future. Otherwise the `async` directive ought not be used.

*tl;dr*
It's not about synchronizing the execution of the callbacks, it's standardizing that they all will return at least tick in the future, and never execute synchronously if they have `async` on the function.